Add parameter validation to the cmac module.

AndrzejKurek opened the pull request:
Summary: Building on #1643, this PR adds parameter validation to the cmac module.

AndrzejKurek commented:
@hanno-arm There already is a dedicated test that checks for null parameters of each function, it just does not use the new parameter validation test error checking scheme. As for the first point, I personally prefer not to double checks, but I understand that if you want to have a uniform behaviour of calling the callback if any input error happens, I should add them here too. I'll wait for opinions from others.

mpg commented:
@AndrzejKurek I think the checks with `CMAC_VALIDATE_RET()` should _replace_ the existing `NULL` checks. See what Hanno did with ChachaPoly in #2275

AndrzejKurek commented on include/mbedtls/cmac.h:
```diff
@@ -96,7 +96,8 @@ int mbedtls_cipher_cmac_starts( mbedtls_cipher_context_t *ctx,
  *
  * \param ctx           The cipher context used for the CMAC operation.
- * \param input         The buffer holding the input data.
+ *                      This must be initialized.
```
I'm not sure if we should duplicate the information that is just above.

AndrzejKurek commented on include/mbedtls/cmac.h:
```diff
@@ -115,7 +116,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
  *                      mbedtls_cipher_cmac_update(), or mbedtls_cipher_free().
  *
  * \param ctx           The cipher context used for the CMAC operation.
+ *                      This must be initialized.
```
I'm not sure if we should duplicate the information that is just above. Come to think of it, I'd even remove it.

AndrzejKurek commented:
@hanno-arm - I've updated this PR. Please re-review.

k-stachowiak commented on tests/suites/test_suite_cmac.function:
```diff
@@ -27,71 +27,56 @@ void mbedtls_cmac_null_args(  )
     /* Test NULL cipher info */
-    TEST_ASSERT( mbedtls_cipher_cmac_update( &ctx, test_data, 16 ) ==
-                                         MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    TEST_INVALID_PARAM( mbedtls_cipher_cmac_update( &ctx, test_data, 16 ) );
```
It appears that the test against the null `cipher_info` validation is only executed for the `..._update()` function and not the others.

k-stachowiak commented on tests/suites/test_suite_cmac.function:
```diff
@@ -27,71 +27,56 @@ void mbedtls_cmac_null_args(  )
 
     cipher_info = mbedtls_cipher_info_from_type( MBEDTLS_CIPHER_AES_128_ECB );
-    TEST_ASSERT( mbedtls_cipher_setup( &ctx, cipher_info ) == 0 );
+    TEST_ASSERT( mbedtls_cipher_setup( &ctx, cipher_info ) == 0 ); 
```
A trailing white space has crept in here.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -97,7 +98,9 @@ int mbedtls_cipher_cmac_starts( mbedtls_cipher_context_t *ctx,
- * \param input         The buffer holding the input data. This must not be \c NULL.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
```
Minor: Use backticks for formatting.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -158,9 +161,12 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
- * \param input         The buffer holding the input data. This must not be \c NULL.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
```
Minor: Use backticks for formatting.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -158,9 +161,12 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
  * \param ilen          The length of the input data.
- * \param output        The buffer for the generic CMAC result. This must not be \c NULL.
+ * \param output        The buffer for the generic CMAC result.
+ *                      This must not be \c NULL.
```
Minor: This too should be of the form "This must be a writable buffer ..."

hanno-arm commented on library/cmac.c:
```diff
@@ -247,9 +252,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    MBEDTLS_CMAC_VALIDATE_RET( ctx != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( ctx->cipher_info != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
It should be `ilen == 0 || input != NULL`.

hanno-arm commented on library/cmac.c:
```diff
@@ -395,8 +401,10 @@ int mbedtls_cipher_cmac( const mbedtls_cipher_info_t *cipher_info,
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    MBEDTLS_CMAC_VALIDATE_RET( cipher_info != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( key != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
It should be `ilen == 0 || input != NULL`.

hanno-arm commented on library/cmac.c:
```diff
@@ -432,8 +440,9 @@ int mbedtls_aes_cmac_prf_128( const unsigned char *key, size_t key_length,
-    if( key == NULL || input == NULL || output == NULL )
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+	MBEDTLS_CMAC_VALIDATE_RET( key != NULL );
+	MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
It should be `ilen == 0 || input != NULL`.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -115,7 +119,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
  *
  * \param ctx           The cipher context used for the CMAC operation.
+ *                      This must be initialized.
  * \param output        The output buffer for the CMAC checksum result.
```
We must document the required size of `output` here. To my understanding, it's the block-size of the underlying cipher.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -151,12 +158,15 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
  * \param ilen          The length of the input data.
  * \param output        The buffer for the generic CMAC result.
```
We must document the required size of `output` here. To my understanding, it's the block-size of the underlying cipher.

hanno-arm commented on include/mbedtls/cmac.h:
```diff
@@ -176,12 +186,12 @@ int mbedtls_cipher_cmac( const mbedtls_cipher_info_t *cipher_info,
- * \param input     The buffer holding the input data.
+ * \param input     The buffer holding the input data. This must not be \c NULL.
  * \param in_len    The length of the input data in Bytes.
  * \param output    The buffer holding the generated 16 Bytes of
```
Suggestion: This must be a writable buffer of size \c 16 Bytes.

sbutcher-arm commented:
PR #1643 has been merged, so removing the `needs: preceding PR` label.

sbutcher-arm commented on library/cmac.c:
```diff
@@ -68,6 +68,10 @@
 
 #if !defined(MBEDTLS_CMAC_ALT) || defined(MBEDTLS_SELF_TEST)
 
+#define MBEDTLS_CMAC_VALIDATE_RET(cond) \
```
These are breaking `check_names.sh` and need to have the `MBEDTLS_` prefix removed.

AndrzejKurek commented:
I've updated the PR with requested changes, fixed a pre-existing documentation bug (wrong error mentioned in the header file), and fixed & extended the tests. @k-stachowiak, @hanno-arm - please re-review.
